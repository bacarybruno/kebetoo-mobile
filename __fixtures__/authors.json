[{
  "uid": "1",
  "email": "bacarybruno@kebetoo.com",
  "displayName": "Bruno Bodian",
  "photoURL": "https://avatars1.githubusercontent.com/u/14147533"
}, {
  "uid": "2",
  "email": "johndoe@kebetoo.com",
  "displayName": "John Doe",
  "photoURL": null
}]