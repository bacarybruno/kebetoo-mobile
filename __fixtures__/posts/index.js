import audio from './audio.json';
import image from './image.json';
import video from './video.json';
import text from './text.json';
import repost from './repost.json';

const posts = {
  audio,
  image,
  video,
  text,
  repost,
};

export const postsList = Object.values(posts);
export default posts;
