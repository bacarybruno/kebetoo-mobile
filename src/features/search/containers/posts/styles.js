import { StyleSheet } from 'react-native';
import { systemWeights } from 'react-native-typography';

import { metrics } from '@app/theme';

export default (colors) => StyleSheet.create({
  wrapper: {
    flex: 1,
    backgroundColor: colors.background,
  },
  paddingHorizontal: {
    paddingHorizontal: metrics.marginHorizontal,
  },
  sectionHeader: {
    marginVertical: metrics.marginVertical / 1.3,
    color: colors.textPrimary,
    ...systemWeights.semibold,
  },
  sectionHeaderLink: {
    color: colors.link,
    ...systemWeights.bold,
  },
  historyHeader: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  flatlistContent: {
    paddingBottom: metrics.tabBarFullHeight - metrics.tabBarHeight,
  },
  noContent: {
    height: '100%',
  },
  icon: {
    marginRight: metrics.spacing.sm,
  },
  container: {
    flexDirection: 'row',
    alignItems: 'center',
  },
});
