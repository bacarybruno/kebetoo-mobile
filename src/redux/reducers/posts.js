import { combineReducers } from 'redux'

import { mergeArrays } from '@app/shared/helpers/object'

import * as types from '../types'

const initialState = {
  posts: [],
}

const posts = (state = initialState.posts, action) => {
  switch (action.type) {
    case types.API_FETCH_POSTS_SUCCESS:
      return mergeArrays(state, action.payload)
    case types.REPLACE_POSTS:
      return action.payload
    default:
      return state
  }
}

export default combineReducers({
  posts,
})
