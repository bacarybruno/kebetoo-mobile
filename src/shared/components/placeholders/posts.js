import React from 'react'
import { View } from 'react-native'
import {
  Placeholder, PlaceholderMedia, PlaceholderLine, Fade,
} from 'rn-placeholder'

const PlaceholderAvatar = () => (
  <PlaceholderMedia isRound style={{ marginRight: 10 }} />
)

const PlaceholderHeader = () => (
  <Placeholder Animation={Fade} Left={PlaceholderAvatar} style={{ marginBottom: 8 }}>
    <PlaceholderLine height={10} width={50} />
    <PlaceholderLine height={10} width={30} />
  </Placeholder>
)

const PlaceholderContent = () => (
  <Placeholder Animation={Fade}>
    <PlaceholderLine height={52} style={{ marginBottom: 8, borderRadius: 5 }} />
  </Placeholder>
)

const PlaceholderReaction = () => (
  <PlaceholderLine width={15} height={20} style={{ marginRight: 10 }} />
)

const PlaceholderReactions = () => (
  <Placeholder style={{ marginBottom: 30 }}>
    <View style={{ flexDirection: 'row' }}>
      <PlaceholderReaction />
      <PlaceholderReaction />
      <PlaceholderReaction />
      <PlaceholderReaction />
    </View>
  </Placeholder>
)

const PostPlaceholder = () => (
  <>
    <PlaceholderHeader />
    <PlaceholderContent />
    <PlaceholderReactions />
  </>
)

export default React.memo(PostPlaceholder)
