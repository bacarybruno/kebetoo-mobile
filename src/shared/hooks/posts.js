import { useCallback } from 'react'

import { getUsers } from '@app/shared/services'

const usePosts = () => {
  const getAuthors = useCallback(async (authorsToFetch) => {
    const data = {}
    if (authorsToFetch.length > 0) {
      const ids = Array.from(new Set(authorsToFetch))
      if (ids.length === 0) return data
      const docs = await getUsers(ids)
      docs.forEach((doc) => {
        data[doc.id] = {
          displayName: doc.displayName,
          photoURL: doc.photoURL,
          id: doc.id,
        }
      })
    }
    return data
  }, [])

  const getRepostAuthors = useCallback(async (p) => {
    let posts = p
    if (!Array.isArray(p)) {
      posts = [p]
    }
    const authorsToFetch = posts.flatMap((post) => {
      if (!post.repost) return []
      return post.repost.author
    })
    const data = await getAuthors(authorsToFetch)
    return data
  }, [getAuthors])

  return {
    getAuthors,
    getRepostAuthors,
  }
}

export default usePosts
