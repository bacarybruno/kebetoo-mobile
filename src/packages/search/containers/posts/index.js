import React, { useEffect, useState, useCallback } from 'react'
import { View, FlatList } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'
import { useIsFocused } from '@react-navigation/native'

import * as api from 'Kebetoo/src/shared/helpers/http'
import Typography, { types as typos } from 'Kebetoo/src/shared/components/typography'
import BasicPost from 'Kebetoo/src/packages/post/containers/basic-post'
import * as types from 'Kebetoo/src/redux/types'
import { recentSearchHistory } from 'Kebetoo/src/redux/selectors'
import Ionicon from 'react-native-vector-icons/Ionicons'
import colors from 'Kebetoo/src/theme/colors'
import strings from 'Kebetoo/src/config/strings'
import usePosts from 'Kebetoo/src/shared/hooks/posts'

import styles from './styles'
import HistoryItem from '../../components/history-item'
import NoResult from '../../components/no-result'

export const SearchHistoryHeader = ({ onClear }) => (
  <View style={[styles.historyHeader, styles.paddingHorizontal]}>
    <View style={{ flexDirection: 'row', alignItems: 'center' }}>
      <Ionicon name="md-time" size={18} style={{ marginRight: 8 }} color={colors.textPrimary} />
      <Typography
        type={typos.headline5}
        style={styles.sectionHeader}
        text={strings.search.recent_searches}
      />
    </View>
    <Typography
      type={typos.textButton}
      style={[styles.sectionHeader, styles.sectionHeaderLink]}
      onPress={onClear}
      text={strings.search.clear_all}
    />
  </View>
)

const SearchPosts = ({ searchQuery, onSearch, onRecentSearch }) => {
  const [posts, setPosts] = useState([])
  const [authors, setAuthors] = useState({})
  const [isLoading, setIsLoading] = useState(false)

  const { posts: postsHistory } = useSelector(recentSearchHistory)
  const dispatch = useDispatch()

  const { getRepostAuthors } = usePosts()

  const isFocused = useIsFocused()

  useEffect(() => {
    const query = searchQuery.trim()
    if (query.length > 0 && isFocused) {
      setIsLoading(true)
      api.searchPosts(query)
        .then((data) => {
          setPosts(data)
          dispatch({ type: types.ADD_POST_HISTORY, payload: query })
        })
        .finally(() => {
          setIsLoading(false)
        })
    } else {
      setPosts([])
    }
  }, [dispatch, onSearch, searchQuery, isFocused])

  useEffect(() => {
    onSearch(isLoading)
  }, [isLoading, onSearch])

  useEffect(() => {
    const fetchRepostAuthors = async () => {
      const data = await getRepostAuthors(posts)
      setAuthors(data)
    }
    fetchRepostAuthors()
  }, [posts, getRepostAuthors])

  const createKey = useCallback((item, index) => `basic-post-${item.id}-${index}`, [])

  const renderBasicPost = useCallback(({ item }) => (
    <BasicPost
      post={item}
      author={item.author}
      originalAuthor={
        item.repost
          ? authors[item.repost.author]
          : item.author
      }
    />
  ), [authors])

  const renderSearchResultsHeader = useCallback(() => (
    <Typography
      type={typos.headline5}
      systemColor={colors.textTertiary}
      style={styles.sectionHeader}
      text={strings.search.results}
    />
  ), [])

  const onClearAllRecentSearches = useCallback(() => {
    dispatch({ type: types.CLEAR_POST_HISTORY })
  }, [dispatch])

  const onClearRecentSearch = useCallback((payload) => {
    dispatch({ type: types.REMOVE_POST_HISTORY, payload })
  }, [dispatch])

  const renderSearchHistoryHeader = useCallback(() => (
    <SearchHistoryHeader onClear={onClearAllRecentSearches} />
  ), [onClearAllRecentSearches])

  const renderSearchHistory = useCallback(({ item }) => (
    <HistoryItem onPress={onRecentSearch} onDelete={onClearRecentSearch} item={item} />
  ), [onClearRecentSearch, onRecentSearch])

  return (
    <View style={styles.wrapper}>
      {searchQuery.length === 0 && (
        <FlatList
          data={postsHistory}
          contentContainerStyle={styles.flatlistContent}
          keyExtractor={createKey}
          ListHeaderComponent={renderSearchHistoryHeader}
          renderItem={renderSearchHistory}
          removeClippedSubviews
        />
      )}
      {posts.length === 0 && searchQuery.length > 0 && !isLoading && (
        <NoResult query={searchQuery} />
      )}
      {posts.length > 0 && (
        <FlatList
          data={posts}
          contentContainerStyle={[styles.flatlistContent, styles.paddingHorizontal]}
          keyExtractor={createKey}
          ListHeaderComponent={renderSearchResultsHeader}
          renderItem={renderBasicPost}
          removeClippedSubviews
        />
      )}
    </View>
  )
}

export default SearchPosts
