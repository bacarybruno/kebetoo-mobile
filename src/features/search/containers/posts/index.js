import React, { useEffect, useState, useCallback } from 'react'
import { View, FlatList } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'
import { useIsFocused } from '@react-navigation/native'
import Ionicon from 'react-native-vector-icons/Ionicons'

import { api } from '@app/shared/services'
import { Typography } from '@app/shared/components'
import BasicPost from '@app/features/post/containers/basic-post'
import * as types from '@app/redux/types'
import { recentSearchHistory } from '@app/redux/selectors'
import { strings } from '@app/config'
import { useAppColors, useAppStyles, usePosts } from '@app/shared/hooks'

import createThemedStyles from './styles'
import HistoryItem from '../../components/history-item'
import NoResult from '../../components/no-result'

export const SearchHistoryHeader = ({ onClear }) => {
  const styles = useAppStyles(createThemedStyles)
  const colors = useAppColors()
  return (
    <View style={[styles.historyHeader, styles.paddingHorizontal]}>
      <View style={{ flexDirection: 'row', alignItems: 'center' }}>
        <Ionicon name="md-time" size={18} style={{ marginRight: 8 }} color={colors.textPrimary} />
        <Typography
          type={Typography.types.headline5}
          style={styles.sectionHeader}
          text={strings.search.recent_searches}
        />
      </View>
      <Typography
        type={Typography.types.textButton}
        style={[styles.sectionHeader, styles.sectionHeaderLink]}
        onPress={onClear}
        text={strings.search.clear_all}
      />
    </View>
  )
}

const SearchPosts = ({ searchQuery, onSearch, onRecentSearch }) => {
  const [posts, setPosts] = useState([])
  const [authors, setAuthors] = useState({})
  const [isLoading, setIsLoading] = useState(false)

  const styles = useAppStyles(createThemedStyles)
  const colors = useAppColors()

  const { posts: postsHistory } = useSelector(recentSearchHistory)
  const dispatch = useDispatch()

  const { getRepostAuthors } = usePosts()
  const isFocused = useIsFocused()

  useEffect(() => {
    const query = searchQuery.trim()
    if (query.length > 0 && isFocused) {
      setIsLoading(true)
      api.posts.search(query)
        .then((data) => {
          setPosts(data)
          dispatch({ type: types.ADD_POST_HISTORY, payload: query })
        })
        .finally(() => {
          setIsLoading(false)
        })
    } else {
      setPosts([])
    }
  }, [dispatch, onSearch, searchQuery, isFocused])

  useEffect(() => {
    onSearch(isLoading)
  }, [isLoading, onSearch])

  useEffect(() => {
    const fetchRepostAuthors = async () => {
      const data = await getRepostAuthors(posts)
      setAuthors(data)
    }
    fetchRepostAuthors()
  }, [posts, getRepostAuthors])

  const createKey = useCallback((item, index) => `basic-post-${item.id}-${index}`, [])

  const renderBasicPost = useCallback(({ item }) => (
    <BasicPost
      post={item}
      author={item.author}
      originalAuthor={
        item.repost
          ? authors[item.repost.author]
          : item.author
      }
    />
  ), [authors])

  const renderSearchResultsHeader = useCallback(() => (
    <Typography
      type={Typography.types.headline5}
      systemColor={colors.textTertiary}
      style={styles.sectionHeader}
      text={strings.search.results}
    />
  ), [colors.textTertiary, styles.sectionHeader])

  const onClearAllRecentSearches = useCallback(() => {
    dispatch({ type: types.CLEAR_POST_HISTORY })
  }, [dispatch])

  const onClearRecentSearch = useCallback((payload) => {
    dispatch({ type: types.REMOVE_POST_HISTORY, payload })
  }, [dispatch])

  const renderSearchHistoryHeader = useCallback(() => (
    <SearchHistoryHeader onClear={onClearAllRecentSearches} />
  ), [onClearAllRecentSearches])

  const renderSearchHistory = useCallback(({ item }) => (
    <HistoryItem onPress={onRecentSearch} onDelete={onClearRecentSearch} item={item} />
  ), [onClearRecentSearch, onRecentSearch])

  return (
    <View style={styles.wrapper}>
      {searchQuery.length === 0 && (
        <FlatList
          data={postsHistory}
          contentContainerStyle={styles.flatlistContent}
          keyExtractor={createKey}
          ListHeaderComponent={renderSearchHistoryHeader}
          renderItem={renderSearchHistory}
          removeClippedSubviews
        />
      )}
      {posts.length === 0 && searchQuery.length > 0 && !isLoading && (
        <NoResult query={searchQuery} />
      )}
      {posts.length > 0 && (
        <FlatList
          data={posts}
          contentContainerStyle={[styles.flatlistContent, styles.paddingHorizontal]}
          keyExtractor={createKey}
          ListHeaderComponent={renderSearchResultsHeader}
          renderItem={renderBasicPost}
          removeClippedSubviews
        />
      )}
    </View>
  )
}

export default SearchPosts
