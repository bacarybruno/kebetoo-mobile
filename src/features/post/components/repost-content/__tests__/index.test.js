import setupTest from '@app/config/jest-setup';
import posts from '@fixtures/posts';
import authors from '@fixtures/authors';

import RepostContent from '../index';

const givenRepostContent = setupTest(RepostContent)({
  post: posts.repost,
  originalAuthor: authors[0],
  style: { elevation: 1 },
  onPress: jest.fn(),
});

it('renders RepostContent', () => {
  const { wrapper } = givenRepostContent();
  expect(wrapper.toJSON()).toMatchSnapshot();
});
