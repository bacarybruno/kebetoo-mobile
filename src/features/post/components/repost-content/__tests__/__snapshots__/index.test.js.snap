// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders RepostContent 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "elevation": 1,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      style={
        Array [
          Object {
            "borderColor": "rgba(229, 229, 234, 255)",
            "borderLeftWidth": 2,
            "marginVertical": 8,
            "minHeight": 52,
            "paddingLeft": 8,
          },
          false,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "marginBottom": 40,
            },
            Object {
              "marginBottom": 0,
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginBottom": 8,
                },
                Object {
                  "marginBottom": 4,
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <View
                style={
                  Object {
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "height": 30,
                        "width": 30,
                      },
                      Object {
                        "height": 24,
                        "width": 24,
                      },
                      undefined,
                    ]
                  }
                >
                  <Image
                    source={
                      Object {
                        "uri": "https://avatars1.githubusercontent.com/u/14147533",
                      }
                    }
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "rgba(0, 122, 255, 255)",
                          "borderRadius": 1000,
                          "flex": 1,
                          "justifyContent": "center",
                        },
                        Object {
                          "borderRadius": 12,
                        },
                      ]
                    }
                    testID="image-avatar-https://avatars1.githubusercontent.com/u/14147533"
                  />
                </View>
              </View>
              <View
                style={
                  Array [
                    Object {
                      "justifyContent": "space-between",
                      "marginLeft": 8,
                    },
                    Object {
                      "height": 24,
                    },
                    Object {
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontSize": 16,
                        "fontWeight": "400",
                        "letterSpacing": -0.32,
                        "lineHeight": 21,
                      },
                      Array [
                        Object {
                          "color": "rgba(0,0,0,0.87)",
                        },
                        Object {
                          "backgroundColor": "transparent",
                          "fontFamily": "System",
                          "fontWeight": "600",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  Bruno Bodian
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "backgroundColor": "#F5F5F5",
                "borderRadius": 5,
                "minHeight": 52,
                "padding": 16,
              },
              false,
              undefined,
              false,
              false,
            ]
          }
        >
          <View
            style={
              Object {
                "opacity": 0,
              }
            }
          >
            <View
              onLayout={[Function]}
            >
              <Text
                numberOfLines={5}
                style={Object {}}
              >
                <Text
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontSize": 15,
                        "fontWeight": "400",
                        "letterSpacing": -0.24,
                        "lineHeight": 20,
                      },
                      Array [
                        Object {
                          "color": "rgba(0,0,0,0.87)",
                        },
                        Object {
                          "backgroundColor": "transparent",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  I'll input the open-source RAM sensor, that should card the FTP protocol!
                </Text>
              </Text>
            </View>
            <View
              onLayout={[Function]}
              style={
                Object {
                  "left": 0,
                  "opacity": 0,
                  "position": "absolute",
                  "top": 0,
                }
              }
            >
              <Text
                style={Object {}}
              >
                <Text
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontSize": 15,
                        "fontWeight": "400",
                        "letterSpacing": -0.24,
                        "lineHeight": 20,
                      },
                      Array [
                        Object {
                          "color": "rgba(0,0,0,0.87)",
                        },
                        Object {
                          "backgroundColor": "transparent",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  I'll input the open-source RAM sensor, that should card the FTP protocol!
                </Text>
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
